magento/magento2
# Use MenuCategoryData to generate array in TopMenu plugin to allow extension via plugins

### Description
Following [discussion on Magento Community Engineering Slack instance](https://magentocommeng.slack.com/archives/CBSL1DF8B/p1592055254155800), it has become clear that the `private` method `Magento\Catalog\Plugin\Block\Topmenu::getCategoryAsArray()` should be abstracted to allow extension via plugin (as inheritance (via preference) is to be discouraged). It appears that an existing class provides most of the functionality required. This pull request changes `getCategoryAsArray()` to use this existing external class (with `public` method) to allow this data to be augmented via plugin on the latter.

### Related Pull Requests
*None known*

### Fixed Issues
*None known*

### Manual testing scenarios
See also discussion on Slack (link above).

1. Attempt to add an item to output of `getCategoryAsArray()` from external code (third party extension)

### Questions or comments
I have not yet adjusted any tests, however this should not result in any change in functionality.

### Contribution checklist (*)

### Resolved issues:
1. [x] resolves magento/magento2#29664: Use MenuCategoryData to generate array in TopMenu plugin to allow extension via plugins
